src/libopensc/pkcs15-cardos.c

```diff
@@ -0,0 +1,177 @@
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "Original Flags: 0x%8.8lx card->flags:0x%8.8lx", passed->used_flags, passed->card_flags);
+
+	if (passed->card_flags) { /* user forced the flags, use them */
+		passed->new_flags = passed->card_flags; /* from card_atr flags */
+	} else {
+
+		for (i = 0, sa = saa[0]; i < SC_MAX_SUPPORTED_ALGORITHMS; i++, sa++) {
+
+			if (sa->reference == 0 && sa->reference == 0 && sa->mechanism == 0
+					&& sa->operations == 0 && sa->algo_ref == 0)
+				break;
+
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "supported_algos[%d] mechamism:0x%8.8x", i, sa->mechanism);
+			switch(sa->mechanism) {
+			case 0x01 :
+				/*
+				 * Card appears to use lower 4 bits of reference as key, and upper
+				 * 4 bits as mech for card.
+				 * Also has a bug if mechanism = 1 (CKM_RSA_PKCS1) and reference 0x10 
+				 * bit is set mechanism should be 3 (CKM_RSA_X_509) 
+				 * correct the mechanism in tokenInfo
+				 */
+				if (sa->reference & 0x10) {
+					sc_log(card->ctx, "Changeing mechanism to CKM_RSA_X_509 based on reference");
+					passed->new_flags |= SC_ALGORITHM_RSA_RAW
+						| SC_ALGORITHM_RSA_PAD_NONE;
+					sa->mechanism = 0x03;
+				} else
+				passed->new_flags |= SC_ALGORITHM_RSA_PAD_PKCS1;
+				break;
+			case 0x03 :
+				passed->new_flags |= SC_ALGORITHM_RSA_RAW
+					| SC_ALGORITHM_RSA_PAD_NONE;
+				break;
+			case 0x06 :
+				passed->new_flags |= SC_ALGORITHM_RSA_HASH_SHA1;
+				break;
+			case 0x1041:
+				passed->ec_flags |= SC_ALGORITHM_ECDSA_RAW;
+				/* no old_ec_flags */
+				/* TODO turn on sizes from  ec curves OIDS */
+				break;
+			default:
+				sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "UNKNOWN MECH: 0x%8.8x", sa->mechanism);
+			}
+
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "New_flags 0x%8.8lx New_ec_flags: 0x%8.8lx",
+				passed->new_flags, passed->ec_flags);
+		}
+
+		if (passed->new_flags == 0) {
+			if (p15card->tokeninfo && p15card->tokeninfo->flags & SC_PKCS15_TOKEN_EID_COMPLIANT) {
+				sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "EID_COMPLIANT flag found");
+				passed->new_flags = (passed->used_flags & ~SC_ALGORITHM_SPECIFIC_FLAGS) | SC_ALGORITHM_RSA_PAD_PKCS1;
+			} else
+				passed->new_flags = passed->used_flags; /* from default cardos_init */
+		}
+	}
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL,"Final New_flags 0x%8.8lx New_ec_flags: 0x%8.8lx", passed->new_flags, passed->ec_flags);
+
+	passed->pass = 2; /* tell card driver to use the new flags */
+	r = sc_card_ctl(p15card->card, SC_CARDCTL_CARDOS_PASS_ALGO_FLAGS, passed);
+
+	free(passed);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+static int
+cardos_pkcs15emu_detect_card(sc_pkcs15_card_t *p15card)
+{
+	if (p15card->card->type <  SC_CARD_TYPE_CARDOS_BASE)
+		return SC_ERROR_WRONG_CARD;
+
+	if (p15card->card->type >= SC_CARD_TYPE_CARDOS_BASE + 1000)
+		return SC_ERROR_WRONG_CARD;
+
+	return SC_SUCCESS;
+}
+
+
+static int
+sc_pkcs15emu_cardos_init(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
+{
```

Comment:
clang-format suggested the following change:


```
-sc_pkcs15emu_cardos_init(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
-{
+sc_pkcs15emu_cardos_init(struct sc_pkcs15_card *p15card, struct sc_aid *aid) {
```

<!-- identifier: clang-format -->
<!-- File: src/libopensc/pkcs15-cardos.c -->